### PR TITLE
[CHERRY-PICK]fix state dict to save/load learning rate scheduler

### PR DESCRIPTION
**Pull request by zhwesky2010**
### PR types

Function optimization
### PR changes

APIs
### Describe

cherry-pick https://github.com/PaddlePaddle/Paddle/pull/25403
==

1.8新增了多种Decay策略，但与之前的策略有些区别，因此 ``save_dygraph/load_dygraph`` 不可用。这里拓展、优化了基类 ``class Optimizer`` 与基类 ``class LearningRateDecay`` 的``state_dict``、``set_dict``函数，并可以适用于全部的Decay策略，包含以下三种：

- 按step更新的学习率策略，例如 NoamDecay 、 PiecewiseDecay 、NaturalExpDecay等，默认保存：
 ``step_num``

- 按epoch更新的学习率策略，例如 StepDecay 、 MultiStepDecay 、 LambdaDecay等，默认保存：
``epoch_num, learning_rate``

- ReduceLROnPlateau策略，该策略为特殊策略，中间变量很多，需保存：
``cooldown_counter, best_loss, num_bad_epochs, epoch_num, learning_rate``

- 如继续新增竞品对齐策略，不走默认的保存的话，仅重写
```
def _state_keys(self):
    self.keys = ['step_num'] 
```
  即可。

### CODE REVIEW

**Review by lanxianghit — APPROVED** [review]

**Review by swtkiwi — APPROVED** [review]
LGTM
